File: structural/dart/flyweight_structured/text_editor.dart
import 'interfaces.dart';

class ConcreteCharacterFormat extends CharacterFormat {
  @override
  final String font;
  @override
  final int size;
  @override
  final String color;
  final bool bold;
  final bool italic;
  
  ConcreteCharacterFormat(this.font, this.size, this.color, this.bold, this.italic);
  
  @override
  void apply(String text) {
    var output = 'Rendering \'$text\' with font: $font, size: $size, color: $color';
    if (bold) output += ', bold';
    if (italic) output += ', italic';
    print(output);
  }
}

class CharacterFormatFactory {
  final Map<String, CharacterFormat> _formats = {};
  
  String _getKey(String font, int size, String color, bool bold, bool italic) {
    return '$font|$size|$color|${bold ? 'B' : ''}|${italic ? 'I' : ''}';
  }
  
  CharacterFormat getFormat(String font, int size, String color, bool bold, bool italic) {
    var key = _getKey(font, size, color, bold, italic);
    
    if (!_formats.containsKey(key)) {
      print('Creating new format: $font ${size}pt');
      _formats[key] = ConcreteCharacterFormat(font, size, color, bold, italic);
    } else {
      print('Reusing existing format: $font ${size}pt');
    }
    
    return _formats[key]!;
  }
  
  int getTotalFormats() {
    return _formats.length;
  }
}

class Character {
  final String symbol;
  final CharacterFormat format;
  
  Character(this.symbol, this.format);
  
  void render() {
    format.apply(symbol);
  }
}

File: structural/typescript/flyweight_structured/text_editor.ts
import { CharacterFormat } from './interfaces';

export class ConcreteCharacterFormat implements CharacterFormat {
  constructor(
    private font: string,
    private size: number,
    private color: string,
    private bold: boolean,
    private italic: boolean
  ) {}
  
  apply(text: string): void {
    let output = `Rendering '${text}' with font: ${this.font}, size: ${this.size}, color: ${this.color}`;
    if (this.bold) output += ', bold';
    if (this.italic) output += ', italic';
    console.log(output);
  }
  
  getFont(): string {
    return this.font;
  }
  
  getSize(): number {
    return this.size;
  }
  
  getColor(): string {
    return this.color;
  }
}

export class CharacterFormatFactory {
  private formats: Map<string, CharacterFormat> = new Map();
  
  private getKey(font: string, size: number, color: string, bold: boolean, italic: boolean): string {
    return `${font}|${size}|${color}|${bold ? 'B' : ''}|${italic ? 'I' : ''}`;
  }
  
  getFormat(font: string, size: number, color: string, bold: boolean, italic: boolean): CharacterFormat {
    const key = this.getKey(font, size, color, bold, italic);
    
    if (!this.formats.has(key)) {
      console.log(`Creating new format: ${font} ${size}pt`);
      this.formats.set(key, new ConcreteCharacterFormat(font, size, color, bold, italic));
    } else {
      console.log(`Reusing existing format: ${font} ${size}pt`);
    }
    
    return this.formats.get(key)!;
  }
  
  getTotalFormats(): number {
    return this.formats.size;
  }
}

export class Character {
  constructor(
    private symbol: string,
    private format: CharacterFormat
  ) {}
  
  render(): void {
    this.format.apply(this.symbol);
  }
}
